codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,294 @@
+        return true;
+    }
+
+    private boolean decodeInline(ByteBuf in, List<Object> out) throws Exception {
+        ByteBuf lineBytes = readLine(in);
+        if (lineBytes == null) {
+            if (in.readableBytes() > maxInlineMessageLength) {
+                throw new RedisCodecException("length: " + in.readableBytes() +
+                                              " (expected: <= " + maxInlineMessageLength + ")");
+            }
+            return false;
+        }
+        out.add(newInlineRedisMessage(type, lineBytes));
+        resetDecoder();
+        return true;
+    }
+
+    private boolean decodeLength(ByteBuf in, List<Object> out) throws Exception {
+        ByteBuf lineByteBuf = readLine(in);
+        if (lineByteBuf == null) {
+            return false;
+        }
+        final long length = parseRedisNumber(lineByteBuf);
+        if (length < RedisConstants.NULL_VALUE) {
+            throw new RedisCodecException("length: " + length + " (expected: >= -1)");
+        }
+        switch (type) {
+        case ARRAY_HEADER:
+            out.add(new ArrayHeaderRedisMessage(length));
+            resetDecoder();
+            return true;
+        case BULK_STRING:
+            if (length > RedisConstants.REDIS_MESSAGE_MAX_LENGTH) {
+                throw new RedisCodecException("length: " + length + " (expected: <= " +
+                                              RedisConstants.REDIS_MESSAGE_MAX_LENGTH + ")");
+            }
+            remainingBulkLength = (int) length; // range(int) is already checked.
+            out.add(new BulkStringHeaderRedisMessage(remainingBulkLength));
+            return decodeBulkString(remainingBulkLength, in, out);
+        default:
+            throw new RedisCodecException("bad type: " + type);
+        }
+    }
+
+    private boolean decodeBulkString(int length, ByteBuf in, List<Object> out) throws Exception {
+        switch (length) {
+        case RedisConstants.NULL_VALUE: // $-1\r\n
+            out.add(FullBulkStringRedisMessage.NULL_INSTANCE);
+            resetDecoder();
+            return true;
+        case 0:
+            state = State.DECODE_BULK_STRING_EOL;
+            return decodeBulkStringEndOfLine(in, out);
+        default: // expectedBulkLength is always positive.
+            state = State.DECODE_BULK_STRING_CONTENT;
+            return decodeBulkStringContent(in, out);
+        }
+    }
+
+    // $0\r\n <here> \r\n
+    private boolean decodeBulkStringEndOfLine(ByteBuf in, List<Object> out) throws Exception {
+        if (in.readableBytes() < RedisConstants.EOL_LENGTH) {
+            return false;
+        }
+        readEndOfLine(in);
+        out.add(FullBulkStringRedisMessage.EMPTY_INSTANCE);
+        resetDecoder();
+        return true;
+    }
+
+    // ${expectedBulkLength}\r\n <here> {data...}\r\n
+    private boolean decodeBulkStringContent(ByteBuf in, List<Object> out) throws Exception {
+        if (!in.isReadable()) {
```

Comment:
since we check `isReadable` multiple times in this method consider just using `int readableBytes = in.readableBytes()`.